auth_service/auth_adapter/core/auth.py

```
@@ -18,121 +18,207 @@
     access_token: Optional[str], key=None
 ) -> Optional[dict[str, Any]]:
-    """Decode and verify the given JSON Web Token."""
+    """Decode and validate the given JSON Web Token."""
     if not access_token:
         return None
     if not key:
-        key = signing_keys.external_jwks
+        key = jwt_config.external_jwks
         if not key:
-            log.debug("No external signing key, cannot verify token")
+            log.debug("No external signing key, cannot validate token.")
             return None
-    jws_token = jws.JWS()
     try:
-        jws_token.deserialize(access_token, key=key)
-        payload = json.loads(jws_token.payload.decode("UTF-8"))
-        return payload
-    except jws.InvalidJWSObject:
-        log.debug("Invalid access token format")
-    except jws.InvalidJWSSignature:
-        log.debug("Invalid access token signature")
-    except jws.JWKeyNotFound:
-        log.debug("Signature key for access token not found")
-    except UnicodeDecodeError:
-        log.debug("Access token payload has invalid encoding")
-    except json.JSONDecodeError:
-        log.debug("Access token payload is not valid JSON")
+        token = jwt.JWT(
+            jwt=access_token,
+            key=key,
+            algs=jwt_config.external_algs,
+            check_claims=jwt_config.check_claims,
+            expected_type="JWS",
+        )
+        claims = json.loads(token.claims)
+        # in addition to JWT.validate() which checks whether claims exist,
+        # we also make sure that the copied claims are not null or empty strings
+        if not claims["sub"]:
+            raise ValueError("The subject claim is missing.")
+        for claim in jwt_config.copied_claims:
+            if not claims[claim]:
+                raise ValueError(f"Missing value for {claim} claim.")
+        return claims
+    except Exception as exc:  # pylint:disable=broad-except
+        log.debug("Cannot validate external token: %s", exc)
```

For these very generic exceptions, I would just let the service fail.